analyzer/src/main/kotlin/PackageManager.kt

```diff
@@ -141,20 +141,22 @@ abstract class PackageManager(
         }
 
         /**
-         * Enrich a package's VCS information with information deduced from the package's VCS URL or an optional
-         * homepage URL.
+         * Enrich a package's VCS information with information deduced from the package's VCS URL or a list of fallback
+         * URLs.
          *
          * @param vcsFromPackage The [VcsInfo] of a [Package].
-         * @param homepageUrl The URL to the homepage of a [Package], if any.
+         * @param fallbackUrls The list of alternative URLs to to use as fallback for determining the [VcsInfo].
          */
-        fun processPackageVcs(vcsFromPackage: VcsInfo, homepageUrl: String = ""): VcsInfo {
+        fun processPackageVcs(vcsFromPackage: VcsInfo, fallbackUrls: List<String> = emptyList()): VcsInfo {
             val normalizedVcsFromPackage = vcsFromPackage.normalize()
 
-            val normalizedUrl = normalizedVcsFromPackage.url.takeIf {
-                it.isNotEmpty()
-            } ?: normalizeVcsUrl(homepageUrl).takeIf {
-                VersionControlSystem.forUrl(it) != null
-            }.orEmpty()
+            val normalizedUrl = normalizedVcsFromPackage.url.takeIf { it.isNotEmpty() }
+                ?: fallbackUrls
+                    .asSequence()
+                    .map { normalizeVcsUrl(it) }
+                    .filterNot { VersionControlSystem.forUrl(it) == null }
```

Comment:
If we now pass URLs for browsing repositories as fallback URLs, I wonder whether we should simply drop this filter step. It was originally meant for keeping only URLs that actually point to VCS repos, i.e. URLs that work as-is for cloning. This means browsable URLs like https://github.com/babel/babel/blob/v6.26.0/packages/babel-cli/ would be filtered out, as you cannot clone from them. However, the call to `VcsHost.toVcsInfo(normalizedUrl)` below would turn such URLs into VcsInfo that *can* be used for cloning. So I guess we should keep such URLs, or?

If you remove this filter step, it should probably be done in a separate commit to be able to provide a proper rationale in the commit message.

Reply:
I tried that yesterday but then a lot of invalid URLs pass, because in most cases the fallback URLs really point to the homepages and not the VCS. The resulting `vcsProcessed` also have type `NONE`. I think what we would need is something that is not as strict as `VersionControlSystem.forUrl` and not as lenient as `VcsHost.toVcsInfo`. But because this requires more time for testing and implementation I won't look into it now.

Reply:
In our Maven+Gradle+Sbt test projects of all URLs that would pass if we remove this check only two are actually pointing to a VCS, and these are links that point to SourceForge CVS repositories, so I'm not sure it would be worth the effort:
```
http://kxml.cvs.sourceforge.net/viewvc/kxml/kxml2/
http://junit.cvs.sourceforge.net/junit/
```

Reply:
Maybe the solution (for later) is to perform the `VersionControlSystem.forUrl(it) == null` check on `vcsFromUrl.url` instead *after* the call to `VcsHost.toVcsInfo()`?

Reply:
I'm not sure, they both try to parse random URLs into valid VCS URLs, the result should not depend on the order they are called, instead they should be merged somehow so that there is a single function that we can call. And then there is also `normalizeVcsUrl` which is called before the other functions, but also inside `VcsHost.toVcsInfo`, I think this needs some cleanup to make it easier to use.

Reply:
Ideally all of this would be done `VcsInfo.normalize` or a similar `VcsInfo.process`. Btw, `VcsInfo.normalize` has the problem that `normalizeVcsUrl` potentially detects the `VcsType` but it does not get updated in `VcsInfo`.

Reply:
> they both try to parse random URLs into valid VCS URLs

If by "both" you mean `VersionControlSystem.forUrl()` and `VcsHost.toVcsInfo()`, then please note that `VersionControlSystem.forUrl()` does not actually *parse* anything. It simply leverages the underlying VCS implementation to check whether the URL can be cloned from. That is, it potentially requires network connection, making it relatively expensive to call (unless its result is already cached in the map).

`VcsHost.toVcsInfo()` on the other hand just statically parses / decomposes the URL and works completely offline.